### ISSUE TITLE
Compiled FlexAttention is nondeterministic on ROCm (gfx950)

### ISSUE THREAD

**Issue by rishisinhanj** [pytorch/torchtitan#3960]
## Summary

On ROCm (gfx950 / MI350X), `torch.compile(flex_attention)` gives **different results run-to-run on bit-identical inputs**, even with `torch.use_deterministic_algorithms(True)`. Eager `flex_attention` is bitwise exact. This breaks the deterministic loss check on the `8 GPU Feature Tests` ROCm leg (exposed when the llama3 debug model switched its default attention from SDPA to FlexAttention). Only deterministic mode is affected.

## Root cause

The chained `tl.dot(..., input_precision="ieee")` accumulation lowers to MFMA on the AMD backend, where the reduction is combined across wavefronts in a **run-varying order**. fp addition is non-associative, so the order change flips the last bits -- fp32 is exact, LSE is exact, `dv` is exact, and `dq`/`dk` drift (they route through the ill-conditioned `ds = p*(dp - Di)` subtraction, which amplifies the ULP noise). This is an upstream Triton/AMD bug, isolated in **ROCm/triton#953**.

## Repro / verified data

Torch-only. Runs compiled vs eager fwd+bwd on bit-identical inputs and reports the max run-to-run diff (nonzero == nondeterministic):

```python
import os, torch
from torch.nn.attention.flex_attention import create_block_mask, flex_attention

os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"
torch.use_deterministic_algorithms(True)

B, H, S, D = 8, 16, 2048, 16
dt, dev = torch.bfloat16, "cuda"
NUM_WARPS = 0  # 0 = autotuned (2 for this shape); set 8 to see it go deterministic

def causal(b, h, q, kv):
    return q >= kv

bm = create_block_mask(causal, B, None, S, S, device=dev)
g = torch.Generator(device=dev).manual_seed(42)
mk = lambda: torch.randn(B, H, S, D, dtype=dt, device=dev, generator=g, requires_grad=True)
q, k, v = mk(), mk(), mk()
go = torch.randn(B, H, S, D, dtype=dt, device=dev, generator=g)
ko = {"num_warps": NUM_WARPS} if NUM_WARPS else {}

def run_once(fn):
    for t in (q, k, v):
        t.grad = None
    out = fn(q, k, v, block_mask=bm, kernel_options=ko)
    out.backward(go)
    return out.clone(), q.grad.clone(), k.grad.clone(), v.grad.clone()

def check(name, fn, iters=10):
    ref = run_once(fn)
    worst = [0.0] * 4
    for _ in range(iters - 1):
        cur = run_once(fn)
        worst = [max(w, (a.float() - b.float()).abs().max().item())
                 for w, a, b in zip(worst, ref, cur)]
    print(f"[{name}] out={worst[0]:.2e} dq={worst[1]:.2e} dk={worst[2]:.2e} dv={worst[3]:.2e} "
          f"-> {'NON-DET' if any(worst) else 'deterministic'}")

check("compiled", torch.compile(flex_attention))
check("eager", flex_attention)
```

On the CI nightly (torch 2.14.0.dev20260719+rocm7.2 / triton-rocm 3.8.0), causal `B=8 H=16 S=2048 D=16` bf16:

| num_warps | out(fwd) | dq | dk | dv |
|---|---|---|---|---|
| 2 | 5.41e-1 | 3.7e-2 | 2.9e-2 | 0 |
| 4 | 2.24e-1 | 7.5e-2 | 2.3e-2 | 0 |
| 8 | 0 | 0 | 0 | 0 |
| eager | 0 | 0 | 0 | 0 |

Requires the ROCm nightly on gfx950; the torch 2.10 release is deterministic. `--B 1 --H 1`, `S=512`, and `float32` are also exact.

## Current fix (not long-term)

**#3896 falls back to eager flex under deterministic mode on ROCm.** This makes CI deterministic now, but we do **not** want it as the long-term solution -- it disables compiled FlexAttention for all ROCm runs, not just CI. (`num_warps>=8` also removes the drift but is fragile: `num_warps=4` was deterministic on triton 3.7.0 and drifts on 3.8.0, so it shifts across versions.)

## Long-term

- [ ] ROCm/triton#953: deterministic `tl.dot` accumulation in the AMD backend (the real fix)
- [ ] PyTorch Inductor: Triton flex path should honor `use_deterministic_algorithms` (pin a config or raise, like the FLASH backend already does) instead of silently drifting
- [ ] Remove the eager-only ROCm stopgap once the above lands

Refs: #3896, ROCm/triton#953
